### PR TITLE
Add public IP address resource

### PR DESCRIPTION
**Pull request by kybeak**
For more information:
https://microsoft.sharepoint.com/:w:/t/ASZ/Ecza9JKQ_pNFn7wxf65IRnkBZXnMuaRcLpZ-um53UvabWA?e=hkgvtW&nav=eyJoIjoiMjMzMzEyNjk4In0&isSPOFile=1&clickparams=eyJBcHBOYW1lIjoiVGVhbXMtRGVza3RvcCIsIkFwcFZlcnNpb24iOiI0OS8yNDExMDExNTcwOSIsIkhhc0ZlZGVyYXRlZFVzZXIiOmZhbHNlfQ%3D%3D

### CODE REVIEW

**Comment by kybeak** [comment]
MOC 2411.2 work item: 
https://msazure.visualstudio.com/One/_queries/query/c9eeab12-b023-4982-baa3-f089f57ffd52/
Feature work item: 
https://msazure.visualstudio.com/One/_workitems/edit/27098761

**Comment by kybeak** [comment]
moc-cicd-daily-scheduled result:
https://msazure.visualstudio.com/msk8s/_build/results?buildId=108897919&view=results

File: services/network/publicipaddress/wssd.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+func getPublicIPAddressPrecheckResponse(response *wssdcloudnetwork.PublicIPAddressPrecheckResponse) (bool, error) {
+	result := response.GetResult().GetValue()
+	if !result {
+		return result, errors.New(response.GetError())
+	}
+	return result, nil
+}
+
+// Validate inputs
+func ValidateInputs(ctx context.Context, group, name string, inputPip *network.PublicIPAddress) error {
```
[the remark this answers is not in the document]
Comment (kybeak) [reply services/network/publicipaddress/wssd.go:136]: You mean setting default if not set?

File: services/network/publicipaddress/wssd.go [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+		return errors.Wrapf(errors.InvalidInput, "Public IP address name not specified")
+	}
+
+	if inputPip == nil || inputPip.PublicIPAddressPropertiesFormat == nil {
+		return errors.Wrapf(errors.InvalidConfiguration, "Missing public IP address properties")
+	}
+
+	if inputPip.PublicIPAddressPropertiesFormat.IdleTimeoutInMinutes == nil {
+		// Set default if not set
+		*inputPip.PublicIPAddressPropertiesFormat.IdleTimeoutInMinutes = 4
```
[the remark this answers is not in the document]
Comment (kybeak) [reply services/network/publicipaddress/wssd.go:152]: Would you please elaborate where it gets nilptr?

File: services/network/networkinterface/networkinterface.go [hunk 3]
```diff
@@ -136,6 +136,14 @@ func getWssdNetworkInterfaceIPConfig(ipConfig *network.InterfaceIPConfiguration,
 			},
 		}
 	}
+	if ipConfig.PublicIPAddress != nil {
+		wssdipconfig.PublicIPAddressRef = &wssdcommonproto.PublicIPAddressReference{
+			ResourceRef: &wssdcommonproto.ResourceReference{
+				Name: *ipConfig.PublicIPAddress.ID,
```
[the remark this answers is not in the document]
Comment (kybeak) [reply services/network/networkinterface/networkinterface.go:142]: That is right, but I validate the ID at wssdclouldagent. 

	// Validate public ip address reference inputs if SDN is enabled
	if err = c.ensurePublicIPAddressRef(ctx, vnicDef); err != nil {
		return
	}

File: services/network/publicipaddress/publicipaddress.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	"github.com/microsoft/moc/pkg/status"
+	"github.com/microsoft/moc/pkg/tags"
+	wssdcloudnetwork "github.com/microsoft/moc/rpc/cloudagent/network"
+	wssdcommonproto "github.com/microsoft/moc/rpc/common"
+)
+
+// GetWssdPublicIPAddress converts our internal representation of a PublicIPAddress (network.PublicIPAddress) to the cloud public IP address protobuf used by wssdcloudagent (wssdnetwork.PublicIPAddress)
+func getWssdPublicIPAddress(networkPip *network.PublicIPAddress, group string) (wssdCloudPip *wssdcloudnetwork.PublicIPAddress, err error) {
+
+	if len(group) == 0 {
```
[the remark this answers is not in the document]
Comment (kybeak) [reply services/network/publicipaddress/publicipaddress.go:19]: ValidateInputs is for create and update operation. This is for delete operation which doesn't pass public ip resource, except name and group.

File: services/network/publicipaddress/wssd.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kybeak) [reply services/network/publicipaddress/wssd.go]: Rename the function 'validateInputsAndSetDefaults'

File: services/network/publicipaddress/publicipaddress.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kybeak) [reply services/network/publicipaddress/publicipaddress.go]: Sure, I will fix it.

File: services/network/network.go [hunk 5]
```diff
@@ -684,7 +684,7 @@ type InterfaceIPConfigurationPropertiesFormat struct {
 	// Primary - Gets whether this is a primary customer address on the network interface.
 	Primary *bool `json:"primary,omitempty"`
 	// PublicIPAddress - Public IP address bound to the IP configuration.
-	PublicIPAddress *PublicIPAddress `json:"publicIPAddress,omitempty"`
+	PublicIPAddress *SubResource `json:"publicIPAddress,omitempty"`
```
Comment (nijosmsft) [inline services/network/network.go:687]: >SubResource [](http://example.com/codeflow?start=18&length=11)

can you please point me to the azure models.go? the one i am looking at, its not SubResource

File: services/network/network.go [hunk 6]
```diff
@@ -961,24 +961,20 @@ type IPTag struct {
...
-	DNSSettings *PublicIPAddressDNSSettings `json:"dnsSettings,omitempty"`
-	// IPTags - The list of tags associated with the public IP address.
-	IPTags *[]IPTag `json:"ipTags,omitempty"`
-	// IPAddress - The IP address associated with the public IP address resource.
-	IPAddress *string `json:"ipAddress,omitempty"`
-	// PublicIPPrefix - The Public IP Prefix this Public IP Address should be allocated from.
-	PublicIPPrefix *SubResource `json:"publicIPPrefix,omitempty"`
 	// IdleTimeoutInMinutes - The idle timeout of the public IP address.
-	IdleTimeoutInMinutes *int32 `json:"idleTimeoutInMinutes,omitempty"`
+	IdleTimeoutInMinutes *uint32 `json:"idleTimeoutInMinutes,omitempty"`
```
Comment (nijosmsft) [inline services/network/network.go:977]: >uint32 [](http://example.com/codeflow?start=23&length=6)

this too, can you please make sure we are aligned with azure?

File: services/network/network.go [hunk 7]
```diff
@@ -992,14 +988,14 @@ type PublicIPAddress struct {
...
-	// Zones - A list of availability zones denoting the IP allocated for the resource needs to come from.
-	Zones *[]string `json:"zones,omitempty"`
 	// ID - Resource ID.
 	ID *string `json:"id,omitempty"`
 	// Name - READ-ONLY; Resource name.
 	Name *string `json:"name,omitempty"`
 	// Type - READ-ONLY; Resource type.
 	Type *string `json:"type,omitempty"`
+	// Version
+	Version *string `json:"version,omitempty"`
```
Comment (nijosmsft) [inline services/network/network.go:998]: >Version *string `json:"version,omitempty"` [](http://example.com/codeflow?start=1&length=42)

this too, as far as i see its PublicIPAddressPropertiesFormat->PublicIPAddressVersion

File: services/network/network.go (hunk 6 above)
Comment (kybeak) [reply services/network/network.go:977]: int32 to unit 32. There is no way for IdleTimeoutInMinutes can be negative.

File: services/network/network.go (hunk 7 above)
Comment (kybeak) [reply services/network/network.go:998]: Version is not the IP version. I believe you or your team asked to have resource version for all. This was specifically asked. 

PublicIPAddressPropertiesFormat contains the pip ip address version.
type PublicIPAddressPropertiesFormat struct {
......
    // PublicIPAddressVersion - The public IP address version. Possible values include: 'IPv4', 'IPv6'
    PublicIPAddressVersion IPVersion `json:"publicIPAddressVersion,omitempty"`

**Comment by kybeak** [comment]
Test results: Main and V2 
https://msazure.visualstudio.com/msk8s/_build/results?buildId=110273373&view=results
https://msazure.visualstudio.com/msk8s/_build/results?buildId=110273471&view=results

**Comment by kybeak** [comment]
https://msazure.visualstudio.com/msk8s/_build/results?buildId=110403677&view=results
https://msazure.visualstudio.com/msk8s/_build/results?buildId=110403205&view=results
https://msazure.visualstudio.com/msk8s/_build/results?buildId=110357396&view=results

File: go.mod [hunk 8]
```diff
@@ -62,6 +62,7 @@ replace (
 	github.com/dgrijalva/jwt-go => github.com/golang-jwt/jwt/v4 v4.1.0
 	github.com/gogo/protobuf => github.com/gogo/protobuf v1.3.2
 	github.com/hashicorp/go-retryablehttp => github.com/hashicorp/go-retryablehttp v0.7.7
+	github.com/microsoft/moc => github.com/microsoft/moc v0.25.2-0.20241212023151-07f45e6a42e2
```
Comment (ganeshamohana) [inline go.mod:65]: Is this going to be changed to an official version before merging?

File: pkg/client/network.go [hunk 9]
```diff
@@ -79,3 +79,14 @@ func GetNetworkSecurityGroupClient(serverAddress *string, authorizer auth.Author
 
 	return network_pb.NewNetworkSecurityGroupAgentClient(conn), nil
 }
+
+// GetPublicIPAddressAgentClient creates a new PublicIPAddressAgentClient using the provided server address and authorizer.
+// It establishes a client connection and returns the PublicIPAddressAgentClient.
+func GetPublicIPAddressAgentClient(serverAddress *string, authorizer auth.Authorizer) (network_pb.PublicIPAddressAgentClient, error) {
+	conn, err := getClientConnection(serverAddress, authorizer)
+	if err != nil {
+		log.Fatalf("Unable to get PublicIPAddressAgentClient. Failed to dial: %v", err)
```
Comment (ganeshamohana) [inline pkg/client/network.go:88]: Do we not need a return statement here with error handling?

File: services/network/publicipaddress/wssd.go [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+		return nil, err
+	}
+
+	response, err := c.PublicIPAddressAgentClient.Invoke(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	pips, err := c.getPublicIPAddressesFromResponse(response)
+	if err != nil {
```
Comment (ganeshamohana) [inline services/network/publicipaddress/wssd.go:77]: Can we add check to validate the array size before we access the index below?

File: services/network/publicipaddress/wssd.go [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+
+	if pip == nil {
+		return nil, errors.InvalidInput
+	}
+
+	request := &wssdcloudnetwork.PublicIPAddressRequest{
+		OperationType:     opType,
+		PublicIPAddresses: []*wssdcloudnetwork.PublicIPAddress{},
+	}
+	var err error
```
Comment (ganeshamohana) [inline services/network/publicipaddress/wssd.go:227]: why do we need this explicit declaration?

File: go.mod (hunk 8 above)
Comment (kybeak) [reply go.mod:65]: Yes, it must be updated later once the moc repo is approved and make official build.

**Comment by kybeak** [comment]
MOC cicd daily scheduled and SDN test results (sdn_cluster_static_e2e and sdn_cni_cluster_static_e2e):
https://msazure.visualstudio.com/msk8s/_build/results?buildId=110550067&view=results
https://msazure.visualstudio.com/msk8s/_build/results?buildId=110553199&view=results

**Review by ganeshamohana — DISMISSED** [review]

**Review by nijosmsft — APPROVED** [review]

**Review by nikohobart — APPROVED** [review]
Re-approving in Ganesh's place since he's OOF.

**Pull request merged**